### ISSUE TITLE
Unknown websocket-related segfault

### ISSUE THREAD

**Issue by cohoe** [jocosocial/swiftarr#53]
After HAL opened up websockets, the app started crashing way more often. No explicit logs were generated of why. Chall added a ton of debug statements to try and find what happened but we were unsuccessful. This led to users seeing a "Bad Gateway" error as the app restarted.

**Comment by cohoe** [comment]
Was able to reproduce this with a websocket testing script I found. Had to fill in a session cookie and adjust the WS url to a valid seamail conversation.

```
#!/usr/bin/env python

import websocket

# https://pypi.org/project/websocket-client/

def on_message(ws, message):
    print(message)

def on_error(ws, error):
    print(error)

def on_close(ws, close_status_code, close_msg):
    print("### closed ###")

def on_open(ws):
    print("Opened connection")

if __name__ == "__main__":
    websocket.enableTrace(True)
    cookie = "swiftarr_session=Wcw4u2/NSdKmgYWzg9MjWroEEyNvGRg0uhzPjPz8BUM="
    ws = websocket.create_connection(
                              "ws://localhost:8081/fez/9B57F214-7269-4952-B40B-B126CB16FFB6/socket",
                              on_open=on_open,
                              on_message=on_message,
                              on_error=on_error,
                              on_close=on_close,
                              cookie=cookie)
    print(ws.recv())
```

Basically grab a valid cookie and seamail ID, throw that in a loop: `for i in $(seq 1 300); do ./ws.py &; done`

Then `pkill -f ws.py` when you're done or your CPU will cry.
For linux, to enable coredumps: `sudo sysctl -w kernel.core_pattern="/tmp/%e_core_dump.%p"`
That'll spit out a file akin to: `/tmp/NIO-ELT-0-#1_core_dump.433172`

My gut tells me it's Linux/NIO/Swift/something interaction that's at fault. But I have little evidence to back up that claim.

```
Process 440089 stopped
* thread #8, name = 'NIO-ELT-0-#6', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
    frame #0: 0x00007ffff74106fe libswiftCore.so`$sSD8_VariantV6lookupyq_SgxF + 14
libswiftCore.so`$sSD8_VariantV6lookupyq_SgxF:
->  0x7ffff74106fe <+14>: cmpq   $0x0, 0x10(%rsi)
    0x7ffff7410703 <+19>: je     0x7ffff7410745            ; <+85>
    0x7ffff7410705 <+21>: movq   %rsi, %r13
    0x7ffff7410708 <+24>: movq   %rdx, %rsi
```
via `lldb .build/x86_64-unknown-linux-gnu/debug/Run`

**Comment by cohoe** [comment]
Got squirrely and did some further investigation. Will have  PR soon with a Locust test that reliably reproduces this issue based on the code above. When running on an x86_64 Mac Mini I get a variety of different crash points. I don't think the actual HTTP request is relevant since the logger is likely catching up.

```
[ INFO ] GET /api/v3/fez/8E910516-A83E-4D4A-80C9-EDDF467E4467 -> ok [request-id: E56204B3-C7F6-4972-90F7-29EE72B7D662]
Vapor/Core.swift:94: Fatal error: Core not configured. Configure with app.core.initialize()
zsh: illegal hardware instruction  swift run

[ INFO ] GET /fez/2E51B145-A47B-4774-934F-6313C82CD4F5/socket -> unauthorized [request-id: 1B92CC99-0ADE-43B5-B865-5CF9248CA320]
AsyncKit/EventLoopGroupConnectionPool.swift:296: Assertion failed: ConnectionPool.shutdown() was not called before deinit.
zsh: illegal hardware instruction  swift run

[ INFO ] GET /fez/85E9A123-317E-443F-8A72-07319BA386BB/socket -> switchingProtocols [request-id: 75C92119-9A73-427F-B0DB-3BE0079CD6E8]
Vapor/Application+Clients.swift:47: Fatal error: Clients not initialized. Initialize with app.clients.initialize()
zsh: illegal hardware instruction  swift run

[ INFO ] POST /api/v3/auth/login -> ok [request-id: A425D1E9-D8AD-4983-9FB2-F36D8E9F5DE9]
AsyncKit/EventLoopGroupConnectionPool.swift:296: Assertion failed: ConnectionPool.shutdown() was not called before deinit.
zsh: illegal hardware instruction  swift run

[ INFO ] POST /api/v3/auth/login -> ok [request-id: 23D32972-5B19-48E2-9D02-2D22A08DE089]
Redis/RedisStorage.swift:51: Fatal error: No redis found for id default, or the app may not have finished booting. Also, the eventLoop must be from Application's EventLoopGroup.
zsh: illegal hardware instruction  swift run

[ INFO ] GET /api/v3/fez/D6005383-07B4-4320-BAD2-450932A6E8FF -> ok [request-id: B30A0FAE-B23D-432C-A12A-EAF1FFAFCD53]
AsyncKit/EventLoopGroupConnectionPool.swift:296: Assertion failed: ConnectionPool.shutdown() was not called before deinit.
zsh: illegal hardware instruction  swift run
```

**Comment by cohoe** [comment]
Apparently the default `maximumActiveConnections` in `RedisConfiguration.PoolOptions` is `2` which is pretty tiny. I played with bumping this to various values including 20, 300, 900 (I run my locust with 300 clients). Under the latter value I end up with different errors such as:

```
[ INFO ] GET /fez/A3CFE2B2-9271-428F-B723-7876DE786707/socket -> unauthorized [request-id: 31DC7448-C277-40FB-985E-0ACDE5611882]
Redis/RedisStorage.swift:51: Fatal error: No redis found for id default, or the app may not have finished booting. Also, the eventLoop must be from Application's EventLoopGroup.

[ INFO ] GET /fez/148E4596-3DD2-462C-BBB1-913BB1735277/socket -> switchingProtocols [request-id: BEDC9999-01D8-4FF4-B52B-B4440BDCEF46]
Redis/RedisStorage.swift:51: Fatal error: No redis found for id default, or the app may not have finished booting. Also, the eventLoop must be from Application's EventLoopGroup.

[ INFO ] POST /api/v3/auth/login -> ok [request-id: 95128172-0A1F-4EC8-ACDD-5F2EA8D5CF69]
Vapor/Core.swift:94: Fatal error: Core not configured. Configure with app.core.initialize()

[ INFO ] GET /api/v3/fez/14A3EC6A-34C6-412E-A9C1-7E2879B1C022 -> ok [request-id: 99C535E4-5F79-48CF-A086-2F0D3A3E9BF2]
AsyncHTTPClient/HTTPClient.swift:132: Fatal error: Client not shut down before the deinit. Please call client.syncShutdown() when no longer needed. Otherwise memory will leak.
```

My gut tells me that we're dealing with a Redis connection pool issue. I wonder what the relationship between WebSocket and Redis connection is since we store each session in Redis.

**Comment by cohoe** [comment]
I take it all back. Got some more thread dumps on linux. They all start the same way.

```
* thread #1, name = 'Run', stop reason = signal SIGSEGV
  * frame #0: 0x00007fd5eb1c69ee libswiftCore.so`$sSD8_VariantV6lookupyq_SgxF + 14
    frame #1: 0x00007fd5eb3ffdf6 libswiftCore.so`$sSDyq_Sgxcig + 6
    frame #2: 0x0000557fc002bf50 Run`$s5Vapor7StorageV3gety5ValueQzSgxmAA0B3KeyRzlF at Storage.swift:44:39
    frame #3: 0x0000557fc002bdac Run`$s5Vapor7StorageVy5ValueQzSgxmcAA0B3KeyRzluig at Storage.swift:30:18
    frame #4: 0x000055aa7a0136be Run`$s5Vapor11ApplicationC6FluentEADV7storageAE7StorageCvg at FluentProvider.swift:125:44                                                                                                                    
    frame #5: 0x000055aa7a012fc5 Run`$s5Vapor11ApplicationC6FluentE9databases0C3Kit9DatabasesCvg at FluentProvider.swift:43:21                                                                                                                
    frame #6: 0x000055aa7a012afb Run`$s5Vapor7RequestC6FluentE2dby0C3Kit8Database_pAF0F2IDVSgF at FluentProvider.swift:11:14                                                                                                                  
    frame #7: 0x000055aa7a012a70 Run`$s5Vapor7RequestC6FluentE2db0C3Kit8Database_pvg at FluentProvider.swift:6:14

* thread #1, name = 'Run', stop reason = signal SIGSEGV
  * frame #0: 0x00007f68bdc969ee libswiftCore.so`$sSD8_VariantV6lookupyq_SgxF + 14
    frame #1: 0x00007f68bdecfdf6 libswiftCore.so`$sSDyq_Sgxcig + 6
    frame #2: 0x00005587536c4f50 Run`$s5Vapor7StorageV3gety5ValueQzSgxmAA0B3KeyRzlF at Storage.swift:44:39
    frame #3: 0x00005587536c4dac Run`$s5Vapor7StorageVy5ValueQzSgxmcAA0B3KeyRzluig at Storage.swift:30:18
    frame #4: 0x00005587535bd815 Run`$s5Vapor11ApplicationC4HTTPV6ClientV6shared15AsyncHTTPClient0G0Cvg at Application+HTTP+Client.swift:23:55
    frame #5: 0x00005587535bd57d Run`$s5Vapor11ApplicationC7ClientsV8ProviderV4httpAGvgZyACcfU_AA6Client_pACcfU_ at Application+HTTP+Client.swift:5:32
    frame #6: 0x000055875356e10a Run`$s5Vapor11ApplicationC6clientAA6Client_pvg at Application+Clients.swift:10:16
    frame #7: 0x000055875357896a Run`$s5Vapor7RequestC6clientAA6Client_pvg at Request+Client.swift:3:26

* thread #1, name = 'Run', stop reason = signal SIGSEGV
  * frame #0: 0x00007f29b2dd09ee libswiftCore.so`$sSD8_VariantV6lookupyq_SgxF + 14
    frame #1: 0x00007f29b3009df6 libswiftCore.so`$sSDyq_Sgxcig + 6
    frame #2: 0x0000559a2c729f50 Run`$s5Vapor7StorageV3gety5ValueQzSgxmAA0B3KeyRzlF at Storage.swift:44:39
    frame #3: 0x0000559a2c729dac Run`$s5Vapor7StorageVy5ValueQzSgxmcAA0B3KeyRzluig at Storage.swift:30:18
    frame #4: 0x0000559a2c5374e1 Run`$s5Vapor11ApplicationC5RedisE12redisStorageAD0cE0Cvg at RedisStorage.swift:8:39
    frame #5: 0x0000559a2c527345 Run`$s5Vapor11ApplicationC5RedisEADV4pool3for9RediStack0C14ConnectionPoolC7NIOCore9EventLoop_p_tF at Application+Redis.swift:17:30
    frame #6: 0x0000559a2c53afbc Run`$s5Vapor7RequestC5RedisEADV4send7command4with7NIOCore15EventLoopFutureCy9RediStack9RESPValueOGSS_SayANGtF at Request+Redis.swift:31:14
    frame #7: 0x0000559a2c53bcc5 Run`$s5Vapor7RequestC5RedisEADV9RediStack0C6ClientAdfGP4send7command4with7NIOCore15EventLoopFutureCyAF9RESPValueOGSS_SayAPGtFTW at <compiler-generated>:0

* thread #1, name = 'Run', stop reason = signal SIGSEGV
  * frame #0: 0x00007fec204109ee libswiftCore.so`$sSD8_VariantV6lookupyq_SgxF + 14
    frame #1: 0x00007fec20649df6 libswiftCore.so`$sSDyq_Sgxcig + 6
    frame #2: 0x00005630e32e7f50 Run`$s5Vapor7StorageV3gety5ValueQzSgxmAA0B3KeyRzlF at Storage.swift:44:39
    frame #3: 0x00005630e32e7dac Run`$s5Vapor7StorageVy5ValueQzSgxmcAA0B3KeyRzluig at Storage.swift:30:18
    frame #4: 0x00005630e281000d Run`$s5Vapor11ApplicationC6FluentEADV10PaginationV13pageSizeLimitSiSgvg at Fluent+Pagination.swift:55:43
    frame #5: 0x00005630e2814e60 Run`$s5Vapor7RequestC6FluentE2dby0C3Kit8Database_pAF0F2IDVSgF at FluentProvider.swift:17:159
    frame #6: 0x00005630e2814a70 Run`$s5Vapor7RequestC6FluentE2db0C3Kit8Database_pvg at FluentProvider.swift:6:14
    frame #7: 0x00005630e19253e2 Run`$s3App13FezControllerV10fezHandleryAA0B4DataV5Vapor7RequestCYaKFTQ0_ at FezController.swift:236:65
```

So I dug into [Storage.swift](https://github.com/vapor/vapor/blob/main/Sources/Vapor/Utilities/Storage.swift) a bit. And in searching the issue tracker: https://github.com/vapor/vapor/issues/2330

That is VERY on point!

**Comment by cohoe** [comment]
Load test put in via https://github.com/jocosocial/swiftarr/pull/79

**Comment by cohoe** [comment]
Well, good news! We no longer seem to be suffering this bug. Not sure what changed since August (`Storage.swift` in Vapor hasn't) but something under the hood has. At elevated socket levels (>700-ish) we start to get Redis connection pool errors:
```
[ WARNING ] The operation could not be completed. (RediStack.RedisConnectionPoolError error 1.) [request-id: 63F589A2-CFAC-41A1-9E8D-EEC668AD7CD4]
```
By increasing the `.maximumActiveConnections()` from the current default of 2 to 4 (in `configure.swift`) I was able to make these mostly go away for a Locust user count of 1000. Also need to fix the `SeamailWebsocketUser` test to not close connections at the end, to ensure they stick around long enough to simulate a relevant real-world case. Anywho....

I was able to crank the websocket connection count up to north of 1500. No crashes. This bodes very well. I'd feel better if we knew what exactly fixed this glitch.
